### PR TITLE
SWATCH-2572 Configure Ansible-aaS

### PR DESCRIPTION
**Pull request by kartikshahc**
Jira issue: SWATCH-2572 and related card in Ansible side: AAP-24829

## Description

This configuration is to enable Ansible with level 2 attribute. We have a following card to start using Level 2 attribute SWATCH-2656

## Testing

IQE Test MR: 

### Steps 1 when we receive valid product tag

1. Kafka producer to service-instance-ingress:
`{"sla": "Premium", "usage": "Production", "org_id": "3340851", "event_id": "668800cf-8cb6-4066-b95f-4a2ebc52018b", "timestamp": "2024-06-10T10:00:00Z", "conversion": false, "event_type": "managed-host", "expiration": "2024-06-10T11:00:00Z", "instance_id": "d147ddf2-be4a-4a59-acf7-7f222758b47c", "product_tag": ["ansible-aap-managed"], "record_date": "2024-06-10T17:20:21.570490791Z", "display_name": "automation__cluster_d147ddf2-be4a-4a59-acf7-7f222758b47c", "event_source": "Premium", "measurements": [{"uom": "Managed-nodes", "value": 4.0, "metric_id": "Managed-nodes"}], "service_type": "Ansible Managed Node"}
`
### Verification

1. It should persist the event 
`668800cf-8cb6-4066-b95f-4a2ebc52018b,2024-06-10 10:00:00.000000 +00:00,"{""sla"": ""Premium"", ""usage"": ""Production"", ""org_id"": ""3340851"", ""event_id"": ""668800cf-8cb6-4066-b95f-4a2ebc52018b"", ""timestamp"": ""2024-06-10T10:00:00Z"", ""conversion"": false, ""event_type"": ""managed-host"", ""expiration"": ""2024-06-10T11:00:00Z"", ""instance_id"": ""d147ddf2-be4a-4a59-acf7-7f222758b47c"", ""product_tag"": [""ansible-aap-managed""], ""record_date"": ""2024-06-10T17:20:21.570490791Z"", ""display_name"": ""automation__cluster_d147ddf2-be4a-4a59-acf7-7f222758b47c"", ""event_source"": ""Premium"", ""measurements"": [{""uom"": ""Managed-nodes"", ""value"": 4.0, ""metric_id"": ""Managed-nodes""}], ""service_type"": ""Ansible Managed Node""}",managed-host,Premium,d147ddf2-be4a-4a59-acf7-7f222758b47c,3340851,,2024-06-10 17:20:21.570491 +00:00`

### Steps 1 when we receive invalid product tag

1. Kafka producer to service-instance-ingress:
`{"sla": "Premium", "usage": "Production", "org_id": "3340851", "event_id": "668800cf-8cb6-4066-b95f-4a2ebc52018b", "timestamp": "2024-06-10T10:00:00Z", "conversion": false, "event_type": "managed-host", "expiration": "2024-06-10T11:00:00Z", "instance_id": "d147ddf2-be4a-4a59-acf7-7f222758b47c", "product_tag": ["dummy"], "record_date": "2024-06-10T17:20:21.570490791Z", "display_name": "automation__cluster_d147ddf2-be4a-4a59-acf7-7f222758b47c", "event_source": "Premium", "measurements": [{"uom": "Managed-nodes", "value": 4.0, "metric_id": "Managed-nodes"}], "service_type": "Ansible Managed Node"}`

### Verification
```
2024-06-10 17:32:38,794 [thread=swatch-instance-ingress-0-C-1] [WARN ] [org.candlepin.subscriptions.event.EventController] - Product tag Optional[dummy] is invalid.
2024-06-10 17:32:38,794 [thread=swatch-instance-ingress-0-C-1] [WARN ] [org.candlepin.subscriptions.event.EventController] - SUBSCRIPTIONS3005: An invalid service instance event was encountered and will be skipped. 
```

### CODE REVIEW

File: swatch-product-configuration/src/main/resources/subscription_configs/Ansible/Ansible_Aap_Managed.yaml
Comment (kahowell) [inline swatch-product-configuration/src/main/resources/subscription_configs/Ansible/Ansible_Aap_Managed.yaml:22]:
```suggestion
  - id: Managed-hosts
    awsDimension: managed_hosts
```

**Review by kahowell — CHANGES_REQUESTED** [review]

File: swatch-product-configuration/src/main/resources/subscription_configs/Ansible/Ansible_Aap_Managed.yaml
Comment (kahowell) [inline swatch-product-configuration/src/main/resources/subscription_configs/Ansible/Ansible_Aap_Managed.yaml:1]: Would like to rename to `ansible-aap-managed.yaml`. No reason to keep the filename different than the `id` value.

File: swatch-product-configuration/src/main/resources/subscription_configs/Ansible/Ansible_Aap_Managed.yaml
Comment (kartikshahc) [reply swatch-product-configuration/src/main/resources/subscription_configs/Ansible/Ansible_Aap_Managed.yaml]: Done

Comment (kartikshahc) [reply swatch-product-configuration/src/main/resources/subscription_configs/Ansible/Ansible_Aap_Managed.yaml]: changed to Managed-nodes instead of Managed-hosts

**Review by kahowell — APPROVED** [review]

**Comment by san7ket** [comment]
Events are not persisted for aap.

Test steps:

Check swatch-product-configuration for aap
```
In [5]: config_parser.get_product("ansible-aap-managed")
Out[5]: 
{'platform': 'Ansible',
 'id': 'ansible-aap-managed',
 'variants': [{'tag': 'ansible-aap-managed',
   'productNames': ['Ansible Automation Platform Managed App']}],
 'defaults': {'variant': 'ansible-aap-managed',
  'sla': 'PREMIUM',
  'usage': 'PRODUCTION'},
 'serviceType': 'Ansible Managed Node',
 'contractEnabled': True,
 'metrics': [{'id': 'Managed-nodes', 'awsDimension': 'managed_nodes'}]}
```
Create events using the events api
```
In [9]: app.rhsm_subscriptions.create_mock_payg_cluster(product_id="ansible-aap-managed" ,**{'managed-nodes':4})
2024-06-10 11:54:00.958 [    INFO] [root] [{'event_source': 'Premium', 'event_type': 'managed-host', 'org_id': '3340851', 'product_tag': ['ansible-aap-managed'], 'instance_id': 'd147ddf2-be4a-4a59-acf7-7f222758b47c', 'timestamp': '2024-06-10T10:00:00Z', 'expiration': '2024-06-10T11:00:00Z', 'display_name': 'automation__cluster_d147ddf2-be4a-4a59-acf7-7f222758b47c', 'measurements': [{'value': 4, 'uom': 'Managed-nodes', 'metric_id': 'Managed-nodes'}], 'sla': 'Premium', 'service_type': 'Ansible Managed Node', 'install_uuid': 'fe9c0c8e-aaca-49fc-b85a-e652fcfc3ee3', 'instance_uuid': 'fe9c0c8e-aaca-49fc-b85a-e652fcfc3ee3/wFFybKXnAA', 'usage': 'Production'}]
2024-06-10 11:54:00.959 [    INFO] [root] Adding payg events of org 3340851 for clowder_smoke env ...
2024-06-10 11:54:00.973 [    INFO] [root] POST request response url: http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/rpc/tally/events, status_code: 200
2024-06-10 11:54:00.973 [    INFO] [root] Response: {
    "detail": "Events saved"
}
2024-06-10 11:54:00.973 [    INFO] [root] ansible-aap-managed PAYG Cluster d147ddf2-be4a-4a59-acf7-7f222758b47c created successfully!!
Out[9]: 'd147ddf2-be4a-4a59-acf7-7f222758b47c'
```
**Logs:**
```
2024-06-10 11:52:53,812 [thread=http-nio-8000-exec-9] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
2024-06-10 11:53:01,478 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
2024-06-10 11:53:04,864 [thread=http-nio-8000-exec-3] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
2024-06-10 11:53:08,713 [thread=http-nio-8000-exec-5] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
2024-06-10 11:53:13,171 [thread=http-nio-8000-exec-7] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
2024-06-10 11:53:36,676 [thread=http-nio-8000-exec-9] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
2024-06-10 11:54:00,968 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events requested by user: self
2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [WARN ] [org.candlepin.subscriptions.tally.admin.InternalTallyDataController] self- Invalid event in batch: org.candlepin.subscriptions.json.Event@12903867[eventSource=Premium,eventType=managed-host,orgId=3340851,instanceId=d147ddf2-be4a-4a59-acf7-7f222758b47c,meteringBatchId=<null>,eventId=<null>,serviceType=Ansible Managed Node,timestamp=2024-06-10T10:00Z,recordDate=<null>,expiration=Optional[2024-06-10T11:00Z],displayName=Optional[automation__cluster_d147ddf2-be4a-4a59-acf7-7f222758b47c],inventoryId=<null>,insightsId=<null>,subscriptionManagerId=<null>,correlationIds=<null>,measurements=[org.candlepin.subscriptions.json.Measurement@69d14c85[value=4.0,uom=Managed-nodes,metricId=Managed-nodes]],cloudProvider=<null>,hardwareType=<null>,hypervisorUuid=<null>,productIds=<null>,role=<null>,sla=Premium,usage=Production,billingProvider=<null>,billingAccountId=<null>,azureTenantId=<null>,azureResourceId=<null>,azureSubscriptionId=<null>,productTag=[ansible-aa...
2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.event.EventController] self- matching payg product tags for role=null, productIds=null, productName=null, conversion=false: []
2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.event.EventController] self- event.product_tags=[ansible-aap-managed]
2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [WARN ] [org.candlepin.subscriptions.event.EventController] self- Event doesn't match configured product tags in swatch
```

**Comment by kartikshahc** [comment]
> Events are not persisted for aap.
> 
> Test steps:
> 
> Check swatch-product-configuration for aap
> 
> ```
> In [5]: config_parser.get_product("ansible-aap-managed")
> Out[5]: 
> {'platform': 'Ansible',
>  'id': 'ansible-aap-managed',
>  'variants': [{'tag': 'ansible-aap-managed',
>    'productNames': ['Ansible Automation Platform Managed App']}],
>  'defaults': {'variant': 'ansible-aap-managed',
>   'sla': 'PREMIUM',
>   'usage': 'PRODUCTION'},
>  'serviceType': 'Ansible Managed Node',
>  'contractEnabled': True,
>  'metrics': [{'id': 'Managed-nodes', 'awsDimension': 'managed_nodes'}]}
> ```
> 
> Create events using the events api
> 
> ```
> In [9]: app.rhsm_subscriptions.create_mock_payg_cluster(product_id="ansible-aap-managed" ,**{'managed-nodes':4})
> 2024-06-10 11:54:00.958 [    INFO] [root] [{'event_source': 'Premium', 'event_type': 'managed-host', 'org_id': '3340851', 'product_tag': ['ansible-aap-managed'], 'instance_id': 'd147ddf2-be4a-4a59-acf7-7f222758b47c', 'timestamp': '2024-06-10T10:00:00Z', 'expiration': '2024-06-10T11:00:00Z', 'display_name': 'automation__cluster_d147ddf2-be4a-4a59-acf7-7f222758b47c', 'measurements': [{'value': 4, 'uom': 'Managed-nodes', 'metric_id': 'Managed-nodes'}], 'sla': 'Premium', 'service_type': 'Ansible Managed Node', 'install_uuid': 'fe9c0c8e-aaca-49fc-b85a-e652fcfc3ee3', 'instance_uuid': 'fe9c0c8e-aaca-49fc-b85a-e652fcfc3ee3/wFFybKXnAA', 'usage': 'Production'}]
> 2024-06-10 11:54:00.959 [    INFO] [root] Adding payg events of org 3340851 for clowder_smoke env ...
> 2024-06-10 11:54:00.973 [    INFO] [root] POST request response url: http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/rpc/tally/events, status_code: 200
> 2024-06-10 11:54:00.973 [    INFO] [root] Response: {
>     "detail": "Events saved"
> }
> 2024-06-10 11:54:00.973 [    INFO] [root] ansible-aap-managed PAYG Cluster d147ddf2-be4a-4a59-acf7-7f222758b47c created successfully!!
> Out[9]: 'd147ddf2-be4a-4a59-acf7-7f222758b47c'
> ```
> 
> **Logs:**
> 
> ```
> 2024-06-10 11:52:53,812 [thread=http-nio-8000-exec-9] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
> 2024-06-10 11:53:01,478 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
> 2024-06-10 11:53:04,864 [thread=http-nio-8000-exec-3] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
> 2024-06-10 11:53:08,713 [thread=http-nio-8000-exec-5] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
> 2024-06-10 11:53:13,171 [thread=http-nio-8000-exec-7] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
> 2024-06-10 11:53:36,676 [thread=http-nio-8000-exec-9] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events/3340851 requested by user: self
> 2024-06-10 11:54:00,968 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events requested by user: self
> 2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [WARN ] [org.candlepin.subscriptions.tally.admin.InternalTallyDataController] self- Invalid event in batch: org.candlepin.subscriptions.json.Event@12903867[eventSource=Premium,eventType=managed-host,orgId=3340851,instanceId=d147ddf2-be4a-4a59-acf7-7f222758b47c,meteringBatchId=<null>,eventId=<null>,serviceType=Ansible Managed Node,timestamp=2024-06-10T10:00Z,recordDate=<null>,expiration=Optional[2024-06-10T11:00Z],displayName=Optional[automation__cluster_d147ddf2-be4a-4a59-acf7-7f222758b47c],inventoryId=<null>,insightsId=<null>,subscriptionManagerId=<null>,correlationIds=<null>,measurements=[org.candlepin.subscriptions.json.Measurement@69d14c85[value=4.0,uom=Managed-nodes,metricId=Managed-nodes]],cloudProvider=<null>,hardwareType=<null>,hypervisorUuid=<null>,productIds=<null>,role=<null>,sla=Premium,usage=Production,billingProvider=<null>,billingAccountId=<null>,azureTenantId=<null>,azureResourceId=<null>,azureSubscriptionId=<null>,productTag=[ansible-aa...
> 2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.event.EventController] self- matching payg product tags for role=null, productIds=null, productName=null, conversion=false: []
> 2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [INFO ] [org.candlepin.subscriptions.event.EventController] self- event.product_tags=[ansible-aap-managed]
> 2024-06-10 11:54:00,970 [thread=http-nio-8000-exec-1] [WARN ] [org.candlepin.subscriptions.event.EventController] self- Event doesn't match configured product tags in swatch
> ```

@san7ket Can you test now?

**Comment by san7ket** [comment]
On adding the events , we get an error 
`HTTPError: 400 Client Error:  for url: http://swatch-tally-service:8000/api/rhsm-subscriptions/v1/internal/rpc/tally/events`

In pod logs:
```
2024-06-12 08:13:33,072 [thread=http-nio-8000-exec-8] [INFO ] [org.candlepin.subscriptions.security.LogPrincipalFilter] self- Internal API: /api/rhsm-subscriptions/v1/internal/rpc/tally/events requested by user: self
2024-06-12 08:13:33,074 [thread=http-nio-8000-exec-8] [WARN ] [org.candlepin.subscriptions.tally.admin.InternalTallyDataController] self- Error parsing request body
2024-06-12 08:13:33,075 [thread=http-nio-8000-exec-8] [WARN ] [org.candlepin.subscriptions.exception.mapper.BaseExceptionMapper] self- SUBSCRIPTIONS1001: Cannot construct instance of `java.util.LinkedHashSet` (although at least one Creator exists): no String-argument constructor/factory method to deserialize from String value ('ansible-aap-managed')
at [Source: REDACTED (`StreamReadFeature.INCLUDE_SOURCE_IN_LOCATION` disabled); line: 1, column: 96] (through reference chain: java.util.ArrayList[0]->org.candlepin.subscriptions.json.Event["product_tag"])
```

Wondering, why the unit tests didn't catch this, or is this something that is dependent on actual deployment of service

**Comment by san7ket** [comment]
@kartikshahc could you please rebase, mostly done with all necessary changes in automation, would verify once and we are good to merge

**Comment by kartikshahc** [comment]
> @kartikshahc could you please rebase, mostly done with all necessary changes in automation, would verify once and we are good to merge

@san7ket Done

**Comment by kartikshahc** [comment]
@kahowell and @san7ket I am done with the changes for this card. Once I get an ack from QE I would like to go ahead and merge it.

**Comment by san7ket** [comment]
No capacity testing is done as part of this card.

- Tested event retention
- Tested Tally
- Instance API

**Pull request merged**
